sig-scalability/slos/pod_startup_latency.md

```diff
@@ -0,0 +1,47 @@
+## Pod startup latency SLI/SLO details
+
+### User stories
+- As a user of vanilla Kubernetes, I want some guarantee how quickly my pods
+will be started.
+
+### Other notes
+- Only schedulable and stateless pods contribute to the SLI:
+  - If there is no space in the cluster to place the pod, there is not much
+    we can do about it (it is task for Cluster Autoscaler which should have
+    separate SLIs/SLOs).
+  - If placing a pod requires preempting other pods, that may heavily depend
+    on the application (e.g. on their graceful termination period). We don't
+    want that to contribute to this SLI.
+  - Mounting disks required by non-stateless pods may potentially also require
+    non-negligible time, not fully dependent on Kubernetes.
+- We are explicitly excluding image pulling from time the SLI. This is
+because it highly depends on locality of the image, image registry performance
+characteristic (e.g. throughput), image size itself, etc. Since we have
+no control over any of those (and all of those would significantly affect SLI)
+we decided to simply exclude it.
+- We are also explicitly excluding time to run init containers, as, again, this
+is heavily application-dependent (and does't depend on Kubernetes itself).
+- The answer to question "when pod should be considered as started" is also
+not obvious. We decided for the semantic of "when all its containers are
```

Comment:
Can you please confirm this? I'm not sure if it's "all containers are reported as running". IIRC it is podrunning phase (which is at least one container running or sth like that).

Reply:
I'm not necessary saying that it's what we measure - I'm saying this is what we SHOULD measure. ;)

Reply:
In that case do we have a rationale why we "should" measure it that way? (because our ideas of "pod started" can be different:)

Reply:
the points below describe that

Reply:
Yes, but I can't find the reasoning for e.g why we should consider "all containers running", but not "at least one container running". Not that the latter is what we [seem to define](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) as "PodRunning" status.

Reply:
The definition isn't fully precise. What it's going to address are cases when there are more container within a pod but some of them already finished (e.g. succeeded).

Looking into kubelet code where this is computed seems to confirm my theory:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1293

So what happens:
- if there are container in waiting state => pod pending
- if there are containers in unkwonwn state => pod pending


Now: why I think we should wait for *all*, not just for some?
The reason is that a Pod is defined as "all its containers". So if 2/3 of containers are running, the pod itself is not running (at least I can imagine situations like that very easily - some functionality may simply be not available).
Now, imagine that one container is simply misconfigured (e.g. has non-existing image), so it will never be started. but all other container are already running. Can we report pod as running? No.

I added a comment with explanation.

Reply:
There are few points on which I disagree:

1. If a pod has a crashlooping container - we shouldn't wait for it for the purpose of our SLI. IMO that's out of the scope of "pod startup" from k8s perspective:
  - we don't know why it's failing and it might be some user-space problem (e.g image doesn't exist, seg-fault in the code, etc)
  - it's super easy to break the SLO with that definition

2. If all containers except one start and let's say they finish instantaneously while the remaining is still starting, your SLI will be waiting indefinitely for "all containers running". While "at least one container running (and all containers created)" will be a bound condition.

3. Your argument around "pod is not functional if only 2/3 of it's containers are running" may or may not be true - and it depends on the user pod in question (e.g some pods may be functional with 2/3 running). And I think that is for readiness gates to check, which again I believe is a user-space problem we shouldn't be including in our SLI.

In general, I believe that from k8s perspective our measurement should actually be limited to container creations (i.e setting up their cgroups, network, namespaces, etc) and we're done - what happens next is probably user-space. I wasn't around back then, but the [current semantic](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1301-L1303) of "PodRunning" was probably a conscious choice for some reason?

cc @yujuhong @freehan (as they may have some useful inputs here)

Reply:
The semantic was chosen as a boundary that most people perceive as running.  In restart never pods, for instance, two related containers may start at different times, and one may complete before the other has started, but the end user perceives start as the moment the first container lights up.

If you are omitting pull latency, then I think this SLI should require all containers to have started because an accidental start linearization regression would have an impact on both services and jobs, and an end user that had 10 containers that started linearly would rightly conclude Kubernetes is “broken” when using multiple containers.

Reply:
> If you are omitting pull latency, then I think this SLI should require all containers to have started because an accidental start linearization regression would have an impact on both services and jobs, and an end user that had 10 containers that started linearly would rightly conclude Kubernetes is “broken” when using multiple containers.

That is exactly my main argument (which apparently I wasn't able to describe as I wanted - thanks @smarterclayton )

Also, to address some of @shyamjvs comments.

> we don't know why it's failing and it might be some user-space problem (e.g image doesn't exist, seg-fault in the code, etc)

That's not true - the definition says that all container started (not that they are still running). If your app crashes, that's fine - it was started and failed, but will already be reported as started (potentially even more than once).

> In general, I believe that from k8s perspective our measurement should actually be limited to container creations

+100 for that and that is exactly what we're doing.
But "container creations" means *all* containers, not some containers.

Reply:
I just realized that i wrote not exactly what i wanted to write :)
So this should be now clarified to capture all the above.

Reply:
> But "container creations" means all containers, not some containers.

So to clarify - I do agree with that we should consider `all container creations` as part of the SLI. What I was against was to consider `all containers running` (which is what seemed to be indicated in your comment above mine) - my reason being that 'container running' is more of a user-space thing than k8s-specific. With that SLI, we would catch the linearization argument Clayton pointed above.

That said, the current formulation lgtm now.